### PR TITLE
Moving files to trash if failed to remove them in windows

### LINKED ISSUE #1594: `conda.install.rm_rf` can't delete old environments in Windows + Python 2.7
This issue hit me when trying to use `--force` option in `conda env create` (I added this in #102), and it revealed a possible problem with how conda handles links in Windows + Python 2.7.

# Symptom

Trying to delete an environment (not active) from within Python fails because the original file being linked is locked by Windows.

# Context

* Windows
* Python 2.7 (`os.islink` does not work here, which might affect the `rm_rf` function from `conda.install`)

# Reproducing

This command line is enough to reproduce the problem:

```bat
$ conda create -n test pyyaml && python -c "import yaml;from conda.install import rm_rf;rm_rf('C:\\Miniconda\\envs\\test')"
Fetching package metadata: ......

#
# To activate this environment, use:
# > activate test
#

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Miniconda\lib\site-packages\conda\install.py", line 204, in rm_rf
    shutil.rmtree(path)
  File "C:\Miniconda\lib\shutil.py", line 247, in rmtree
    rmtree(fullname, ignore_errors, onerror)
  File "C:\Miniconda\lib\shutil.py", line 247, in rmtree
    rmtree(fullname, ignore_errors, onerror)
  File "C:\Miniconda\lib\shutil.py", line 252, in rmtree
    onerror(os.remove, fullname, sys.exc_info())
  File "C:\Miniconda\lib\shutil.py", line 250, in rmtree
    os.remove(fullname)
WindowsError: [Error 5] Acesso negado: 'C:\\Miniconda\\envs\\test\\Lib\\site-packages\\yaml.dll'
```

Note that when I import `pyyaml`, I'm locking file `C:\\Miniconda\\Lib\\site-packages\\yaml.dll` (root environment), but that also prevents me from deleting `C:\\Miniconda\\envs\\test\\Lib\\site-packages\\yaml.dll` (another hard link).

# Solution?

Maybe we should add some improved support for detecting and unlinking hardlinks in Windows with Python 2.7, and handle those cases individually instead of trying to `shutil.rmtree` everything.

Possibly from [jarako.windows](https://bitbucket.org/jaraco/jaraco.windows/src/default/jaraco/windows/filesystem/__init__.py#cl-76) or [ntfs/fs.py](https://github.com/sid0/ntfs/blob/master/ntfsutils/fs.py#L88)

### PR DESCRIPTION
**Pull request by campos-ddc**
This change will attempt to 'fix' issues when removing locked files in Windows.

To avoid that problem, I just used the trash mechanism already implemented in conda, and put it into `rm_rf`.

I slightly broke the api for `move_to_trash, but it doesn't look like the `tempdir` was used anywhere.

Fixes #1594

### CODE REVIEW

**Comment by campos-ddc** [comment]
I also noted that you currently try to remove the trash directory whenever something is linked.

That might still be impossible, if those files are being used by conda itself, so it might be worth considering deleting them as soon as conda starts (before any imports), or even better, run a .bat file after conda ends.

**Comment by asmeurer** [comment]
Trash deletion should probably happen in the move_to_trash function. Every time conda starts is not a good idea because we don't want to do stuff on the filesystem for what otherwise would be a no-op command (like `--help` or `--dry-run`). We played with using bat files after conda ends, but it's messy. The way it is done now is much better.

**Comment by asmeurer** [comment]
I think `tempdir` was just there for debugging purposes.

**Comment by campos-ddc** [comment]
Yea, bats are a mess =/

Well, just be aware that right now some files are never being removed from my trash (`yaml.dll`, `pycosat.pyd`, stuff that conda itself uses)

**Comment by asmeurer** [comment]
Hmm. I think we should catalog which files are doing this and set them as no_link.

**Comment by campos-ddc** [comment]
I got `['_yaml.pyd', 'yaml.dll', 'pycosat.pyd']`. There might be more, but I've never seen any others.

**Comment by campos-ddc** [comment]
There aren't many conda dependencies, you could probably get away with never linking any of them while only costing like 10Mb of space.

**Comment by asmeurer** [comment]
@ilanschnell can we add those three files to no_link on Windows?

**Comment by asmeurer** [comment]
I think we need to have a `delete_trash` call somewhere here (maybe in move_to_trash itself?).

File: conda/install.py [hunk 1]
```diff
@@ -525,28 +542,20 @@ def move_to_trash(prefix, f, tempdir=None):
...
-        if tempdir is None:
-            import tempfile
-            trash_dir = tempfile.mkdtemp(dir=trash_dir)
-        else:
-            trash_dir = join(trash_dir, tempdir)
+        trash_dir = tempfile.mkdtemp(dir=trash_dir)
 
         try:
-            try:
-                os.makedirs(join(trash_dir, dirname(f)))
```
Comment (asmeurer) [inline conda/install.py]: Why did you delete this?

**Comment by asmeurer** [comment]
Also delete_trash itself uses rm_rf, so it trash=False should be used there.

File: conda/install.py (hunk 1 above)
Comment (campos-ddc) [inline conda/install.py]: No longer creating a new directory with `dirname(f)`, I just put the file directory in the tempdir.

**Comment by campos-ddc** [comment]
I did use `trash=False` in `delete_trash` (line 517)

**Comment by campos-ddc** [comment]
So call `delete_trash` right after moving something inside it? Its never going to delete the recent file, but might clear up old stuff.

File: conda/install.py (hunk 1 above)
Comment (asmeurer) [inline conda/install.py]: I guess I did this for debugging purposes (you can easily see where the trash files are coming from). I would keep this here, unless it is causing problems.

**Comment by asmeurer** [comment]
Yeah, delete_trash always deletes "old stuff", basically by design. Right now we kind of call it in some places where it is likely to be called often (both in conda and conda-build), But I'm wondering if just calling it in move_to_trash would be more consistent. I guess technically conda doesn't always delete stuff (e.g., if a new package is installed).

File: conda/install.py (hunk 1 above)
Comment (campos-ddc) [inline conda/install.py]: Ok, how about removing the `prefix, f` parameters, always passing full `path`, and put it in the trash inside a directory that matches the path relative to conda root?

e.g.,

`move_to_trash('c:/miniconda/envs/old_env/lib/site-packages/yaml.dll')`

will move the file to:

`c:/miniconda/pkgs/.trash/tmp19827312/envs/old_env/lib/site-packages/yaml.dll`

Comment (campos-ddc) [inline conda/install.py]: I pushed another commit with the behavior I just described, please take a look at it,

Comment (asmeurer) [inline conda/install.py]: Yes, that sounds good.

File: conda/install.py [hunk 2]
```diff
@@ -504,47 +514,40 @@ def delete_trash(prefix):
...
     from conda import config
 
     for pkg_dir in config.pkgs_dirs:
+        import tempfile
         trash_dir = join(pkg_dir, '.trash')
 
-        try:
+        if not isdir(trash_dir):
             os.makedirs(trash_dir)
-        except OSError as e1:
```
Comment (asmeurer) [inline conda/install.py]: I think this `continue` logic was here to deal with permissions errors.

Comment (campos-ddc) [inline conda/install.py]: Oops, I missed the fact that you just ignored errors when it wasn't a 'file already exists'

File: conda/install.py [hunk 3]
```diff
@@ -497,26 +507,28 @@ def is_linked(prefix, dist):
     except IOError:
         return None
 
-def delete_trash(prefix):
+def delete_trash():
```
Comment (asmeurer) [inline conda/install.py]: Huh, I don't know why that argument was put there. It may need to be kept for backwards compatibility for at least a version because conda-build uses this function.

File: conda/install.py [hunk 4]
```diff
@@ -497,26 +507,28 @@ def is_linked(prefix, dist):
...
         trash_dir = join(pkg_dir, '.trash')
         try:
             log.debug("Trying to delete the trash dir %s" % trash_dir)
-            rm_rf(trash_dir, max_retries=1)
+            rm_rf(trash_dir, max_retries=1, trash=False)
         except OSError as e:
             log.debug("Could not delete the trash dir %s (%s)" % (trash_dir, e))
 
-def move_to_trash(prefix, f, tempdir=None):
+def move_to_trash(path):
```
Comment (asmeurer) [inline conda/install.py]: The same here. conda-build uses this function.

**Comment by campos-ddc** [comment]
Ok, see if this works.

I'd like to amend all changes into a single commit before merging this, just give me the 'ok' and I'll do that.

**Comment by asmeurer** [comment]
I don't know why you feel you need to do that, but I'd rather you didn't.

**Comment by campos-ddc** [comment]
I usually don't like a pull request merging in with 5 commits, 4 of them being fixes made during code review.

Anyway, I won't if you prefer it like this.

**Comment by asmeurer** [comment]
I'm also seeing the `yaml.dll` and `_yaml.pyd` never get deleted. I guess that's not an issue here, though. 

The conda and conda-build tests seem to be working, and I don't see any issues in the code. I wish AppVeyor was working but I don't know how to fix it. 

I'm glad to see this change. I suppose a next step will be replacing instance of `os.unlink` with `rm_rf` when they are found to do the wrong thing on Windows. The trash stuff in conda-build is also probably not necessary any more either.
